extra/modules/greenbids-real-time-data/src/main/java/org/prebid/server/hooks/modules/greenbids/real/time/data/model/data/GreenbidsInferenceData.java

```diff
@@ -0,0 +1,187 @@
+        final List<Imp> imps = bidRequest.getImp();
+
+        return imps.stream()
+                .flatMap(imp -> extractMessagesForImp(
+                        imp,
+                        bidRequest,
+                        greenbidsUserAgent,
+                        hostname,
+                        hourBucket,
+                        minuteQuadrant,
+                        dbReader,
+                        mapper))
+                .collect(Collectors.toList());
+    }
+
+    private static Stream<ThrottlingMessage> extractMessagesForImp(
+            Imp imp,
+            BidRequest bidRequest,
+            GreenbidsUserAgent greenbidsUserAgent,
+            String hostname,
+            Integer hourBucket,
+            Integer minuteQuadrant,
+            DatabaseReader dbReader,
+            ObjectMapper mapper) {
+
+        final String impId = imp.getId();
+        final ObjectNode impExt = imp.getExt();
+        final JsonNode bidderNode = extImpPrebid(impExt.get("prebid"), mapper).getBidder();
+        final String ip = Optional.ofNullable(bidRequest.getDevice())
+                .map(Device::getIp)
+                .orElse(null);
+        final String countryFromIp = getCountrySafely(ip, dbReader);
+        return createThrottlingMessages(
+                bidderNode,
+                impId,
+                greenbidsUserAgent,
+                countryFromIp,
+                hostname,
+                hourBucket,
+                minuteQuadrant).stream();
+    }
+
+    private static String getCountrySafely(String ip, DatabaseReader dbReader) {
+        try {
+            return getCountry(ip, dbReader);
+        } catch (IOException | GeoIp2Exception e) {
+            throw new PreBidException("Failed to get country for IP", e);
+        }
+    }
+
+    private static List<ThrottlingMessage> createThrottlingMessages(
+            JsonNode bidderNode,
+            String impId,
+            GreenbidsUserAgent greenbidsUserAgent,
+            String countryFromIp,
+            String hostname,
+            Integer hourBucket,
+            Integer minuteQuadrant) {
+
+        final List<ThrottlingMessage> throttlingImpMessages = new ArrayList<>();
+        if (bidderNode.isObject()) {
+            final ObjectNode bidders = (ObjectNode) bidderNode;
+            final Iterator<String> fieldNames = bidders.fieldNames();
+            while (fieldNames.hasNext()) {
+                final String bidderName = fieldNames.next();
+                throttlingImpMessages.add(buildThrottlingMessage(
+                        bidderName,
+                        impId,
+                        greenbidsUserAgent,
+                        countryFromIp,
+                        hostname,
+                        hourBucket,
+                        minuteQuadrant));
+            }
+        }
+        return throttlingImpMessages;
+    }
+
+    private static ThrottlingMessage buildThrottlingMessage(
```

Comment:
Here in `buildThrottlingMessage ` I've reformatted browser and device out of builder to avoid nested indentation. IMO we can keep it as separate method to not clutter up the argument of `throttlingImpMessages.add()`. 

```
final String browser = Optional.ofNullable(greenbidsUserAgent)
                .map(GreenbidsUserAgent::getBrowser)
                .orElse(StringUtils.EMPTY);

final String device = Optional.ofNullable(greenbidsUserAgent)
        .map(GreenbidsUserAgent::getDevice)
        .orElse(StringUtils.EMPTY);
```

WDYT?